src/components/sound-picker/sound-picker.tsx

```diff
@@ -9,15 +9,124 @@ import "./sound-picker.scss";
 export interface ISoundPickerProps {
   selectedSound: SoundName
   handleSoundChange?: (event: ChangeEvent<HTMLSelectElement>) => void;
+  onRecordingCompleted?: (audioBuffer: AudioBuffer) => void;
 }
 
 export const SoundPicker = (props: ISoundPickerProps) => {
-  const { selectedSound, handleSoundChange } = props;
+  const { selectedSound, handleSoundChange, onRecordingCompleted } = props;
+
+  // While recording, 'chunks' of audio data are appended here
+  let audioRecordingChunks: BlobPart[] | undefined = [];
```

Comment:
I'm sorry, but I don't believe that that analysis is correct. The `accessRecordingChunks` variable is referenced on line 121, in the `onMicIconClicked()` function. It is set from there, to ensure that when a new recording is started that no previous audio chunks remain in that array. Therefore, the variable cannot exist only inside the `accessRecordingStream()` function, as its scope ends at line 96.

Reply:
Line 121 is only clearing `audioRecordingChunks`. But this has literally no effect in practice. You can try it yourself. If you remove this line, this won't change the app behavior at all. There's actually a bigger design issue.

Basically, you can't keep the state in functional components the way you did with `audioRecordingChunks`. On each render, this variable will be initialized and equal to `[]`. It doesn't matter what you set it to in any of the event handlers (including `onMicIconClicked`). During the next re-render you'll end up with an empty array again.

If `audioRecordingChunks` was meant to hold the state between rerenders, and referenced by multiple event handlers or children components, you should use `useState` or `useRef` (`useRef` would make more sense in this particular case).

However, your code works fine, as the **only** place, where this variable is used indeed, is the `accessRecordingStream` function. And this function is called only once withing the component lifetime when the media recorder is created. Also, it's only media recorder event handlers using this variable. So, because of that, each time when your code is using `audioRecordingChunks`, it's referring to the variable from its closure ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures)). Also, note that you're clearing this variable on line 89, at the end of `onstop` event handler. That's why the recorded sounds are not glued together. And that's why you don't need to do anything in `onMicIconClicked` (which wouldn't have any effect anyway - for example, you can try removing line 89 and leaving 121 => sounds will be merged together).

Using closures instead of `useRef` is fine, it's a nice way to simplify code and keep things local. But I think `audioRecordingChunks` should live in the "nearest" closure, where it's actually used to avoid confusion and possible bugs (similar to one from line 121, even though it was unharmful in this case).

Reply:
Sorry, I have some coding habits that I have to un-learn, in the (React) functional component world!